### ISSUE TITLE
sft_llama2 loss getting lower but model generates worse output

### ISSUE THREAD

**Issue by lebronjamesking** [huggingface/trl#733]
As per title, I was using https://github.com/huggingface/trl/blob/main/examples/research_projects/stack_llama_2/scripts/sft_llama2.py on my own dataset. The loss is gettinng lower, however the model performance is getting worse.

**Comment by lebronjamesking** [comment]
{'loss': 1.0037, 'learning_rate': 1e-05, 'epoch': 0.01}                                                                   
{'loss': 0.7644, 'learning_rate': 1e-05, 'epoch': 0.03}                                                                   
{'loss': 0.7018, 'learning_rate': 1e-05, 'epoch': 0.04}                                                                   
{'loss': 0.6524, 'learning_rate': 1e-05, 'epoch': 0.06}                                                                   
{'loss': 0.613, 'learning_rate': 1e-05, 'epoch': 0.07}                                                                    
{'loss': 0.588, 'learning_rate': 1e-05, 'epoch': 0.08}                                                                    
{'loss': 0.5651, 'learning_rate': 1e-05, 'epoch': 0.1}                                                                    
{'loss': 0.5447, 'learning_rate': 1e-05, 'epoch': 0.11}                                                                   
{'loss': 0.5373, 'learning_rate': 1e-05, 'epoch': 0.12}                                                                   
{'loss': 0.5254, 'learning_rate': 1e-05, 'epoch': 0.14}                                                                   
{'loss': 0.5311, 'learning_rate': 1e-05, 'epoch': 0.15}                                                                   
{'loss': 0.5041, 'learning_rate': 1e-05, 'epoch': 0.17}                                                                   
{'loss': 0.5101, 'learning_rate': 1e-05, 'epoch': 0.18}                                                                   
{'loss': 0.4967, 'learning_rate': 1e-05, 'epoch': 0.19}                                                                   
{'loss': 0.492, 'learning_rate': 1e-05, 'epoch': 0.21}                                                                    
{'loss': 0.4838, 'learning_rate': 1e-05, 'epoch': 0.22}                                                                   
{'loss': 0.4749, 'learning_rate': 1e-05, 'epoch': 0.24}                                                                   
{'loss': 0.4601, 'learning_rate': 1e-05, 'epoch': 0.25}                                                                   
{'loss': 0.4597, 'learning_rate': 1e-05, 'epoch': 0.26}                                                                   
{'loss': 0.4558, 'learning_rate': 1e-05, 'epoch': 0.28}                                                                   
{'loss': 0.4524, 'learning_rate': 1e-05, 'epoch': 0.29}                                                                   
{'loss': 0.4576, 'learning_rate': 1e-05, 'epoch': 0.3}                                                                    
{'loss': 0.4597, 'learning_rate': 1e-05, 'epoch': 0.32}                                                                   
{'loss': 0.4579, 'learning_rate': 1e-05, 'epoch': 0.33}                                                                   
{'loss': 0.4354, 'learning_rate': 1e-05, 'epoch': 0.35}                                                                   
{'loss': 0.45, 'learning_rate': 1e-05, 'epoch': 0.36}                                                                     
{'loss': 0.4544, 'learning_rate': 1e-05, 'epoch': 0.37}                                                                   
{'loss': 0.4408, 'learning_rate': 1e-05, 'epoch': 0.39}                                                                   
{'loss': 0.4317, 'learning_rate': 1e-05, 'epoch': 0.4}                                                                    
{'loss': 0.4205, 'learning_rate': 1e-05, 'epoch': 0.42}                                                                   
{'loss': 0.4153, 'learning_rate': 1e-05, 'epoch': 0.43}                                                                   
{'loss': 0.4394, 'learning_rate': 1e-05, 'epoch': 0.44}                                                                   
{'loss': 0.4405, 'learning_rate': 1e-05, 'epoch': 0.46}                                                                   
{'loss': 0.4334, 'learning_rate': 1e-05, 'epoch': 0.47}                                                                   
{'loss': 0.4226, 'learning_rate': 1e-05, 'epoch': 0.49}                                                                   
{'loss': 0.4274, 'learning_rate': 1e-05, 'epoch': 0.5}                                                                    
{'loss': 0.424, 'learning_rate': 1e-05, 'epoch': 0.51}                                                                    
{'loss': 0.4136, 'learning_rate': 1e-05, 'epoch': 0.53}                                                                   
{'loss': 0.4174, 'learning_rate': 1e-05, 'epoch': 0.54}                                                                   
{'loss': 0.4043, 'learning_rate': 1e-05, 'epoch': 0.55}                                                                   
{'loss': 0.4119, 'learning_rate': 1e-05, 'epoch': 0.57}                                                                   
{'loss': 0.4161, 'learning_rate': 1e-05, 'epoch': 0.58}                                                                   
{'loss': 0.41, 'learning_rate': 1e-05, 'epoch': 0.6}                                                                      
{'loss': 0.4078, 'learning_rate': 1e-05, 'epoch': 0.61}                                                                   
{'loss': 0.4157, 'learning_rate': 1e-05, 'epoch': 0.62}                                                                   
{'loss': 0.4012, 'learning_rate': 1e-05, 'epoch': 0.64}                                                                   
{'loss': 0.4052, 'learning_rate': 1e-05, 'epoch': 0.65}                                                                   
{'loss': 0.4401, 'learning_rate': 1e-05, 'epoch': 0.67}                                                                   
{'loss': 0.4081, 'learning_rate': 1e-05, 'epoch': 0.68}                                                                   
{'loss': 0.3825, 'learning_rate': 1e-05, 'epoch': 0.69}                                                                   
{'loss': 0.372, 'learning_rate': 1e-05, 'epoch': 0.71}                                                                    
{'loss': 0.38, 'learning_rate': 1e-05, 'epoch': 0.72}                                                                     
{'loss': 0.4744, 'learning_rate': 1e-05, 'epoch': 0.73}                                                                   
{'loss': 0.3842, 'learning_rate': 1e-05, 'epoch': 0.75}                                                                   
{'loss': 0.359, 'learning_rate': 1e-05, 'epoch': 0.76}                                                                    
{'loss': 0.3638, 'learning_rate': 1e-05, 'epoch': 0.78}

**Comment by lebronjamesking** [comment]
btw, I am using llama2-hf for sft. The code snippet is very much the same as sft_llama2.py. I load model in torch.dtype=torch.float16 and TrainingArguments as follows:

training_args = TrainingArguments(
    output_dir=script_args.output_dir,
    per_device_train_batch_size=script_args.per_device_train_batch_size,
    gradient_accumulation_steps=script_args.gradient_accumulation_steps,
    per_device_eval_batch_size=script_args.per_device_eval_batch_size,
    learning_rate=script_args.learning_rate,
    logging_steps=script_args.logging_steps,
    max_steps=script_args.max_steps,
    report_to=script_args.log_with,
    save_steps=script_args.save_steps,
    group_by_length=script_args.group_by_length,
    lr_scheduler_type=script_args.lr_scheduler_type,
    warmup_steps=script_args.num_warmup_steps,
    optim=script_args.optimizer_type,
    # tf32=True,
    # bf16=True,
    fp16=True,
    remove_unused_columns=True,
    run_name="sft_llama2",
)

**Comment by younesbelkada** [comment]
For llama2 some users reported that using `tokenizer.padding_side = "right"` and making sure to train in bf16 fixed some training stability issues.

**Comment by lebronjamesking** [comment]
> For llama2 some users reported that using `tokenizer.padding_side = "right"` and making sure to train in bf16 fixed some training stability issues. Perhaps you could try that and let us know how it goes

Many thanks, I am trying it now. Getting back later

**Comment by lebronjamesking** [comment]
I have done 3 epoch of sft training. The performance is reasonable. 
I think bf16 is a minor issue in this setting. However, 

> gradient_accumulation_steps

 seems an important factor. Do you have any recommendation on that parameter. I am using gradient_accumulation_steps=8 right now.

**Comment by lvwerra** [comment]
Gradient accumulation is essentially a helper when you can't fit larger batches into memory. The global batch size is `batch_size*gradient_accumulation*num_gpus` where you can increase the latter without any more memory consumption. Usually we use something 100-500 as the globaly batch size.

 Glad to hear that it works

**Comment by lebronjamesking** [comment]
Cool, I'm closing the issue. Many thanks for the clear explanation.

**Comment by lebronjamesking** [comment]
Thank you very much.

**Comment by lebronjamesking** [comment]
Close.
